doc/man3/flux_future_set_init.adoc

```diff
@@ -0,0 +1,163 @@
+
+
+DESCRIPTION
+-----------
+
+A Flux future represents some activity that may be completed with reactor
+watchers and/or message handlers.  It is intended to be wrapped
+by other classes that perform specific activities, providing a common idiom
+for synchronizing on the result.  Typically such a class provides
+class-specific wrappers for `flux_future_create()` and `flux_future_get()`,
+and implements a class-specific init callback.
+
+`flux_future_set_init()` registers an init callback.  The callback sets
+up class-specific watchers on the reactor to handle asynchronous events.
+The watchers must eventually call `flux_future_set_result()` or
+`flux_future_set_error()` to fulfill the future.  For more information
+about when the initialization callback is invoked, see REACTOR CONTEXTS
+below.
+
+`flux_future_set_result()` assigns an opaque _result_ value to the
+future, with optional destructor.  NULL is a valid result.
+The assignment fulfills the future and triggers synchronization,
+after which `flux_future_get()` can be used to obtain the _result_
+value without blocking.  The destructor is called when the future is
+destroyed.
+
+`flux_future_set_error()` similarly fulfills the future and triggers
+synchronization, but instead of assigning a result, it assigns an
+errno value, after which `flux_future_get()` returns -1 with errno
+set to this value.
+
+`flux_future_aux_set()` attaches an object _aux_, with optional destructor
+_destroy_, to the future under an optional _name_.  `flux_future_aux_get()`
+retrieves an object by _name_.  Destructors are called when the future is
+destroyed.  Objects may be stored anonymously under a NULL _name_ to be
+scheduled for destruction without the option of retrieval.
+
+`flux_future_set_flux()` may be used to associate a Flux broker handle
+with a future.  The handle may be retrieved from within an init callback using
+`flux_future_get_flux()` and used to set up message handlers that
+fulfill the future in the same manner as described for reactor watchers.
+
+Futures may "contain" other futures, to arbitrary depth.  That is, an
+init callback may create futures and use their continuations to fulfill
+the containing future in the same manner as reactor watchers and message
+handlers.
+
+
+REACTOR CONTEXTS
+----------------
+
+Internally, a future can operate in two reactor contexts.  The init
+callback may be called in either or both contexts, depending on which
+synchronization functions are called by the user.
+
+The main reactor context involves the reactor passed to `flux_future_create()`.
+This reactor is expected to be run or re-entered by the user after a
+continuation is set up with `flux_future_then()`, and can process the
+future's watchers in parallel with other watchers registered by the
+application.  The call to `flux_future_then()` triggers the init
+callback in this context.
+
+Alternatively, an internal reactor is created when `flux_future_get()` or
```

Comment:
Yes, but I think it would mean every caller would have to provide that code, which makes it rather complicated to use.  Both paths coexist just fine.  For example, you can register a continuation  and call `get()` before the continuation is called and it will block until the result is available, and if you enter the reactor the continuation will also be called.